allow reupload of license if app is not installed

stefanrepl opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines here:
https://github.com/replicatedhq/kots/blob/main/CONTRIBUTING.md.
2. Ensure you have added appropriate tests for your PR. For more information read here:
https://github.com/replicatedhq/kots/blob/main/CONTRIBUTING.md#testing
3. If the PR is unfinished, please mark it as a draft.
-->

#### What type of PR is this?
type::bug

#### What this PR does / why we need it:
This PR allows a user to reupload a license if the license has not been used for a successful installation.  Makes UX much smoother so the user does not have to manually run a command to remove the app.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes [SC-36475](https://app.shortcut.com/replicated/story/36475/unable-to-re-attempt-uploading-the-license-if-it-fails-to-install-the-first-time)

#### Special notes for your reviewer:
Tested in Okteto, Okteto endpoint for kots web: https://kotsadm-web-stefanrepl.replicated.okteto.dev/. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
-->
```release-note
     Allow user to re-upload license if application is not installed.
```

#### Does this PR require documentation?
<!--
If no, just write "NONE" below.
If yes, link to the related https://github.com/replicatedhq/kots.io documentation PR:
-->

stefanrepl commented on pkg/handlers/license.go:
```diff
@@ -309,11 +310,66 @@ func (h *Handler) UploadNewLicense(w http.ResponseWriter, r *http.Request) {
+		for _, app := range apps {
+			decode := scheme.Codecs.UniversalDeserializer().Decode
+			obj, _, err := decode([]byte(app.License), nil, nil)
+			if err != nil {
```
makes sense, done

sgalsaleh reviewed (changes requested):
This logic needs to be implemented for automated installs as well. Under `pkg/automation/automation.go`. I would move this logic to a separate shared function and use it for both.

sgalsaleh commented on pkg/handlers/license.go:
```diff
@@ -309,11 +310,63 @@ func (h *Handler) UploadNewLicense(w http.ResponseWriter, r *http.Request) {
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to list installed apps"))
+			uploadLicenseResponse.Error = err.Error()
+			JSON(w, 500, uploadLicenseResponse)
```
```suggestion
			JSON(w, http.StatusInternalServerError, uploadLicenseResponse)
```

sgalsaleh commented on pkg/handlers/license.go:
```diff
@@ -309,11 +310,63 @@ func (h *Handler) UploadNewLicense(w http.ResponseWriter, r *http.Request) {
+					if err := store.GetStore().RemoveApp(id); err != nil {
+						logger.Error(errors.Wrap(err, "failed to remove app"))
+						uploadLicenseResponse.Error = err.Error()
+						JSON(w, 500, uploadLicenseResponse)
```
```suggestion
						JSON(w, http.StatusInternalServerError, uploadLicenseResponse)
```

sgalsaleh commented on pkg/handlers/license.go:
```diff
@@ -309,11 +310,63 @@ func (h *Handler) UploadNewLicense(w http.ResponseWriter, r *http.Request) {
+				if id, err := store.GetStore().GetAppIDFromSlug(slugProposal); err == nil && id != "" {
+					if err := store.GetStore().RemoveApp(id); err != nil {
+						logger.Error(errors.Wrap(err, "failed to remove app"))
+						uploadLicenseResponse.Error = err.Error()
```
```suggestion
						uploadLicenseResponse.Error = "Failed to remove existing app record"
```

sgalsaleh replied:
let's not return internal errors stack trace to the user
```suggestion
						uploadLicenseResponse.Error = "Failed to remove existing app record"
```

stefanrepl commented:
> 



> This logic needs to be implemented for automated installs as well. Under `pkg/automation/automation.go`. I would move this logic to a separate shared function and use it for both.

done, moved to a shared function and added its use to automated installs

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -5,18 +5,71 @@ import (
+	}
+
+	// get correct slug for app if it is not installed
+	titleForSlug := strings.Replace(verifiedLicense.Spec.AppSlug, ".", "-", 0)
```
why not just get the "not installed yet" apps list using a new function, and then remove the apps that have a license id that match instead of trying to guess the slug here? https://github.com/replicatedhq/kots/pull/2577/files#diff-43a280cac09b9a704ed4f9c88e893f7f46c25ebcb0b26b6b080b40d0599c5efaR48

stefanrepl replied:
thats a very good suggestion, ill implement that instead, thanks

stefanrepl replied:
Although I will have to implement a "ListNotInstalledApps" function in a few places to satisfy interface specs

sgalsaleh commented on pkg/automation/automation.go:
```diff
@@ -309,8 +312,11 @@ func AirgapInstall(appSlug string, additionalFiles map[string][]byte) error {
-		cleanup(&licenseSecret, verifiedLicense.Spec.AppSlug)
-		return errors.Errorf("license already exists for app %s", verifiedLicense.Spec.AppSlug)
+		resolved, err := kotslicense.ResolveExistingLicense(verifiedLicense)
+		if !resolved || err != nil {
```
let's log the error here as well like we did in the other file?

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,65 @@ import (
+func ResolveExistingLicense(verifiedLicense *kotsv1beta1.License) (bool, error) {
+	installedAppSet := make(map[string]bool)
+
+	apps, err := store.GetStore().ListInstalledApps()
```
why is this still needed?

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,65 @@ import (
+	}
+
+	// build set of installed app slugs using this license
+	for _, app := range apps {
```
the set should be built using the "not installed" apps, no?

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,65 @@ import (
+		}
+	}
+
+	for _, app := range notInstalledApps {
```
`notInstalledApps` should already have the license, no need for listing installed apps

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,65 @@ import (
+		return false, err
+	}
+
+	notInstalledApps, err := store.GetStore().ListNotInstalledApps()
```
let's call it "failed" instead of "not installed"?
```suggestion
	notInstalledApps, err := store.GetStore().ListFailedApps()
```

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,43 @@ import (
+		license := obj.(*kotsv1beta1.License)
+
+		if license.Spec.LicenseID == verifiedLicense.Spec.LicenseID {
+			if id, err := store.GetStore().GetAppIDFromSlug(app.Slug); err == nil && id != "" {
```
you already have `app.ID` no need to fetch it

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,41 @@ import (
+		}
+
+		if err := store.GetStore().RemoveApp(app.ID); err != nil {
+			return false, err
```
```suggestion
			return false, errors.Wrap(err, "failed to remove existing app record")
```

sgalsaleh commented on pkg/license/license.go:
```diff
@@ -10,13 +10,41 @@ import (
 	License      *kotsv1beta1.License
 }
 
+func ResolveExistingLicense(verifiedLicense *kotsv1beta1.License) (bool, error) {
```
let's call it "newLicense" instead just to be clearer in the context of this function?
```suggestion
func ResolveExistingLicense(newLicense *kotsv1beta1.License) (bool, error) {
```
